### ISSUE TITLE
Some analyzer messages contain newlines (leads to confusing output)

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#52201]
The analyzer output looks weird when the error messages contain newlines, like some of our newer error messages do.

For example:
```
   info • Missing type annotation • test.dart:6:3 • always_specify_types
  error • 'Foo<Object?>' doesn't conform to the bound 'Foo<Foo<Object?>>' of the type parameter 'Self' • test.dart:6:3 • type_argument_not_matching_bounds
  error • Couldn't infer type parameter 'Self'.

Tried to infer 'Foo<Object?>' for 'Self' which doesn't work:
  Type parameter 'Self' is declared to extend 'Foo<Self>' producing 'Foo<Foo<Object?>>'.

Consider passing explicit type argument(s) to the generic.

 • test.dart:6:3 • could_not_infer
```

When the compiler outputs error messages with newlines, they are semantically important (because e.g. they include little chevrons pointing to the bad code), but that's not really true of the analyzer messages. For example, the message above is equally readable as:

```
   info • Missing type annotation • test.dart:6:3 • always_specify_types
  error • 'Foo<Object?>' doesn't conform to the bound 'Foo<Foo<Object?>>' of the type parameter 'Self' • test.dart:6:3 • type_argument_not_matching_bounds
  error • Couldn't infer type parameter 'Self'. Tried to infer 'Foo<Object?>' for 'Self' which doesn't work, the type parameter 'Self' is declared to extend 'Foo<Self>' producing 'Foo<Foo<Object?>>'. Consider passing explicit type argument(s) to the generic. • test.dart:6:3 • could_not_infer
```

That still leaves the question of whether the message should end with a period, which most of the messages avoid.

I suppose we could make it all one sentence:

```
   info • Missing type annotation • test.dart:6:3 • always_specify_types
  error • 'Foo<Object?>' doesn't conform to the bound 'Foo<Foo<Object?>>' of the type parameter 'Self' • test.dart:6:3 • type_argument_not_matching_bounds
  error • Couldn't infer type parameter 'Self' (tried to infer 'Foo<Object?>' for 'Self' which doesn't work, the type parameter 'Self' is declared to extend 'Foo<Self>' producing 'Foo<Foo<Object?>>'); consider passing explicit type argument(s) to the generic • test.dart:6:3 • could_not_infer
```

**Comment by Hixie** [comment]
Another example:
```
   info • Missing type annotation • test.dart:13:3 • always_specify_types
  error • 'Null' doesn't conform to the bound 'Foo<Null>' of the type parameter 'T' • test.dart:13:3 • type_argument_not_matching_bounds
  error • Couldn't infer type parameter 'T'.

Tried to infer 'Null' for 'T' which doesn't work:
  Type parameter 'T' is declared to extend 'Foo<T>' producing 'Foo<Null>'.
The type 'Null' was inferred from:
  Parameter 'test' declared as     'T?'
                   but argument is 'Null'.

Consider passing explicit type argument(s) to the generic.

 • test.dart:13:3 • could_not_infer
```

The following is not more readable, but I would argue it's not less readable either and has the advantage of fitting the pattern that the analyzer uses for other messages:

```
   info • Missing type annotation • test.dart:13:3 • always_specify_types
  error • 'Null' doesn't conform to the bound 'Foo<Null>' of the type parameter 'T' • test.dart:13:3 • type_argument_not_matching_bounds
  error • Couldn't infer type parameter 'T' (tried to infer 'Null' for 'T' which doesn't work, because the type parameter 'T' is declared to extend 'Foo<T>' producing 'Foo<Null>', which is not 'Null'; the type 'Null' was inferred from the parameter 'test' declared as 'T?' with an argument of type 'Null'); consider passing explicit type argument(s) to the generic • test.dart:13:3 • could_not_infer
```

**Comment by bwilkerson** [comment]
@InMatrix

While I agree that the newlines make the output somewhat more difficult to read, I'd argue that the lack of newlines makes it harder (for me at least) to see where the basic message ends and where the auxiliary information begins.

> I suppose we could make it all one sentence ...

Whether it's the right thing to do is an interesting question, but what we're currently doing is dividing the messaging into several pieces:
- the problem message (a summary of what's wrong),
- additional details (more information about why it's a problem; currently only used for inference failures),
- context messages (other places in the code that are relevant to the problem and a description of why they're relevant), and
- a correction message (suggestions about how to fix the problem).

This has the advantage that we have the ability to format the information in the IDE and on the command line in different ways to better fit the characteristics of each. (We might not do the best possible job of formatting the information, but the ability is, I think, valuable, and we do make use of it.)

Making these messages a single sentence would cause problems in the IDE where we display different portions of the messaging in different contexts.

> That still leaves the question of whether the message should end with a period, which most of the messages avoid.

Actually, all of our messages have a period.

When the command-line analyzer was written, and it was decided to use a bullet between pieces of information, a decision was made to drop the trailing period because it looked cleaner. At that time we didn't have the additional details to display. The decision to drop the period (and even the choice of how to format the output) might want to be rethought in light of the confusion caused by having multiple sentences to display.

**Comment by Hixie** [comment]
It's possible we need a different rendering for when the message is shown in one-line environments vs environments that can show multiple lines per message.

What does VS Code do for these messages?

(FWIW, I first ran into this because my development environment assumed analyzer messages always fit on one line, which was the original protocol for `flutter analyze --write=foo`, and when it tried to sort the error messages, it ended up scattering each line of these error messages all over the place...)

**Comment by bwilkerson** [comment]
> What does VS Code do for these messages?

In the hover text it preserves the line breaks and indentation, and it looks pretty good. In the problems view we loose the indentation, but preserve the line breaks, so it looks ok.

> ... assumed analyzer messages always fit on one line, which was the original protocol for `flutter analyze --write=foo` ...

I wasn't aware of this contract. We might want to consider making `dart analyze` and `flutter analyze` consistent. In `dart analyze` we have a `--format` flag that let's the user specify the format they want to use. There aren't a lot of options, but one of them is JSON, which is nice for tooling that wants to process the output. I don't know whether that would meet the use cases for `--write`, though.

**Comment by Hixie** [comment]
Yeah I don't think it was ever an explicit contract, it just happened to be that way.

Maybe I should just make the `--write` replace runs of spaces and newlines into a single space character and not really worry about it beyond that.
